src/ossl/aes.rs

```diff
@@ -203,90 +203,60 @@ impl AesOperation {
     fn init_params(mech: &CK_MECHANISM) -> KResult<AesParams> {
         match mech.mechanism {
             CKM_AES_CCM => {
-                if mech.ulParameterLen as usize
-                    != ::std::mem::size_of::<CK_CCM_PARAMS>()
-                {
-                    return err_rv!(CKR_ARGUMENTS_BAD);
-                }
-                let ccm_params = mech.pParameter as *const CK_CCM_PARAMS;
-                let datalen = unsafe { (*ccm_params).ulDataLen as usize };
-                let nonce = unsafe { (*ccm_params).pNonce };
-                let noncelen = unsafe { (*ccm_params).ulNonceLen as usize };
-                let maclen = unsafe { (*ccm_params).ulMACLen as usize };
-                let aad = unsafe { (*ccm_params).pAAD };
-                let aadlen = unsafe { (*ccm_params).ulAADLen as usize };
-                if noncelen < 7 || noncelen > 13 {
+                let params = cast_params!(mech, CK_CCM_PARAMS);
+                if params.ulNonceLen < 7 || params.ulNonceLen > 13 {
                     return err_rv!(CKR_MECHANISM_PARAM_INVALID);
                 }
-                let l = 15 - noncelen;
-                if datalen == 0
-                    || datalen > (1 << (8 * l))
-                    || datalen > (u64::MAX as usize) - maclen
+                let l = 15 - params.ulNonceLen;
+                if params.ulDataLen == 0
+                    || params.ulDataLen > (1 << (8 * l))
+                    || (params.ulDataLen + params.ulMACLen) as u64 > u64::MAX
```

Comment:
ah I messed up the conversion indeed, and this check now just causes a carsh, very nice catch.
Of course the point here is generally moot because CK_ULONG is generally defined as u64 anyway (so we can't really test this), but let me get the check right.